### PR TITLE
MPI datatype trans optimization

### PR DESCRIPTION
**Pull request by markusbattarbee**
Currently we use a large neighbourhood when doing AMR to ensure we have all necessary cells for the vlasov solver. This update ensures the VDFs are transferred only for those cells which are actually needed.

Diffs (using a 100-second 843-timestep 1 AMR level IPshock run):
```
markusb@turso01:/wrk/users/markusb/AMRpaper2$ /proj/markusb/vlsvdiff_DP AMR1_tp/bulk.0000100.vlsv AMR1B2/bulk.0000100.vlsv proton/vg_rho 0
INFO Reading in two files.
Statistics on file: size 6.95303e-310 min = 4.94066e-324 max = 6.95303e-310 average = 6.95303e-310 standard deviation 0
Statistics on file: size 9000 min = 1.00011e+06 max = 2.27098e+06 average = 1.7351e+06 standard deviation 3091.99
The absolute 0-distance between both datasets is 209.829
The relative 0-distance between both datasets is 9.23959e-05
The average-shifted absolute 0-distance between both datasets is 211.274
The average-shifted relative 0-distance between both datasets is 9.30323e-05
The absolute 1-distance between both datasets is 345875
The relative 1-distance between both datasets is 2.21489e-05
The average-shifted absolute 1-distance between both datasets is 345580
The average-shifted relative 1-distance between both datasets is 2.213e-05
The absolute 2-distance between both datasets is 4771.79
The relative 2-distance between both datasets is 2.85836e-05
The average-shifted absolute 2-distance between both datasets is 4769.82
The average-shifted relative 2-distance between both datasets is 2.85718e-05
```

speedup:
```
AMR1_tp 
(TIME) total run time 3951.27 s, total simulated time 100.044 s
Propagate 3798s, imbalance 0.1668, 3.142e+06 Cells/s
translate proton 2634s, imbalance 0.007595

AMR1B2
(TIME) total run time 3854.56 s, total simulated time 100.044 s
Propagate 3704s, imbalance 0.1687, 3.222e+06 Cells/s
translate proton 2555s, imbalance 0.006713
```

### CODE REVIEW

**Comment by markusbattarbee** [comment]
And with `--diff` given to `vslvdiff_DP`:
```

Statistics on file: size 9000 min = 1.00011e+06 max = 2.27098e+06 average = 1.7351e+06 standard deviation 3091.99
Statistics on file: size 9000 min = 1.00011e+06 max = 2.27098e+06 average = 1.7351e+06 standard deviation 3091.99
The absolute 0-distance between both datasets is 209.829
The relative 0-distance between both datasets is 9.23959e-05
The average-shifted absolute 0-distance between both datasets is 211.274
The average-shifted relative 0-distance between both datasets is 9.30323e-05
The absolute 1-distance between both datasets is 345875
The relative 1-distance between both datasets is 2.21489e-05
The average-shifted absolute 1-distance between both datasets is 345580
The average-shifted relative 1-distance between both datasets is 2.213e-05
The absolute 2-distance between both datasets is 4771.79
The relative 2-distance between both datasets is 2.85836e-05
The average-shifted absolute 2-distance between both datasets is 4769.82
The average-shifted relative 2-distance between both datasets is 2.85718e-05
```

**Comment by markusbattarbee** [comment]
I'll also run a more standard testpackage test

**Comment by ykempf** [comment]
Was this SPF? I wouldn't expect diffs here, or am I missing something?

**Comment by ursg** [comment]
Indeed, shouldn't this end up having the same data where it is needed, and only have potential differences due to summation ordering?

**Comment by markusbattarbee** [comment]
This was 843 timesteps instead of the 100 used in Flowthrough_amr so it's possible it is just that, or it's possible there's a bug somewhere. I'll run Flowthrough_amr and post some diff pictures.

**Comment by markusbattarbee** [comment]
Well, something is amiss :)
```
vlsvdiff_DP --diff Flowthrough_amr_dev/bulk.0000001.vlsv Flowthrough_amr_MPItrans/bulk.0000001.vlsv proton/vg_rho 0
INFO Reading in two files.
Statistics on file: size 4664 min = 1.04839e+06 max = 2.15588e+06 average = 1.50635e+06 standard deviation 6801.63
Statistics on file: size 4664 min = 1.04839e+06 max = 2.15897e+06 average = 1.50638e+06 standard deviation 6801.22
The absolute 0-distance between both datasets is 9766.62
The relative 0-distance between both datasets is 0.00453022
The average-shifted absolute 0-distance between both datasets is 9803.78
The average-shifted relative 0-distance between both datasets is 0.00454745
The absolute 1-distance between both datasets is 304585
The relative 1-distance between both datasets is 4.33536e-05
The average-shifted absolute 1-distance between both datasets is 463004
The average-shifted relative 1-distance between both datasets is 6.59024e-05
The absolute 2-distance between both datasets is 32922.3
The relative 2-distance between both datasets is 0.000305823
The average-shifted absolute 2-distance between both datasets is 32824.4
The average-shifted relative 2-distance between both datasets is 0.000304913
```
diff proton/vg_rho 0 in y-slice:
AMR_y_map_DIFF_proton_vg_rho_0000001


ranks:
AMR_y_map_vg_rank_0000001

**Comment by ykempf** [comment]
Run 2 processes?

**Comment by markusbattarbee** [comment]
With just two tasks it looks even weirder:
diff proton/vg_rho 0:
AMR1_y_map_DIFF_proton_vg_rho_0000001

ranks:
AMR_y_map_vg_rank_0000001

**Comment by markusbattarbee** [comment]
Some plots to verify the flags:
dark and pale regions are different ranks (2 processes)
purple means the flag is active, check the pseudocolor description to verify which direction is plotted
So in the relevant direction, there ordering must be:
pale grey cells <-> two pale purple cells  <-> two dark purple cells  <-> dark grey cells

These all look correct to me.

Screenshot 2021-02-25 at 16 41 34
Screenshot 2021-02-25 at 16 41 24
Screenshot 2021-02-25 at 16 41 07
Screenshot 2021-02-25 at 16 40 57

**Comment by markusbattarbee** [comment]
Hmpf. I would like to understand why this isn't working as intended, but at least with relatively small vspaces the time spent in transfer-stencil-data-x is only of the order of 2-5 percent of translate, so this won't cause a huge increase even if it worked. Still would be nice to understand this better.

**Comment by markusbattarbee** [comment]
Verified that after 1 step only, diffs are still on a numerical level. At subsequent steps, they slowly grow (Still below 1e-7 at 5 steps). Turning FS and ACC off did not fix things.

File: grid.cpp [hunk 1]
```diff
@@ -566,8 +567,14 @@ void balanceLoad(dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry>& mpiGrid, S
    recalculateLocalCellsCache();
    getObjectWrapper().meshData.reallocate();
    cells = mpiGrid.get_cells();
+   #pragma parallel for
```
Comment (ykempf) [inline grid.cpp:570]: Should that not include an `omp`?

Comment (markusbattarbee) [reply grid.cpp:570]: whoops :)

**Comment by markusbattarbee** [comment]
So I tried adding the
`updateRemoteVelocityBlockLists(mpiGrid,popID,VLASOV_SOLVER_Y_NEIGHBORHOOD_ID);`
commands to dev and testing that, but after 99 timesteps, diffs are still big - so it's apparently not that dev would require that update.

**Comment by markusbattarbee** [comment]
Flowthrough_amr testpackage test on Vorna:
```
$/vlsvdiff_DP --diff Flowthrough_amr_dev_2tasks/bulk.0000001.vlsv Flowthrough_amr_Mpitrans_2tasks/bulk.0000proton/vg_rho 0
INFO Reading in two files.
Statistics on file: size 4664 min = 1.04839e+06 max = 2.15588e+06 average = 1.50635e+06 standard deviation 6801.63
Statistics on file: size 4664 min = 1.04839e+06 max = 2.15588e+06 average = 1.50635e+06 standard deviation 6801.63
The absolute 0-distance between both datasets is 0.000807128
The relative 0-distance between both datasets is 3.74384e-10
The average-shifted absolute 0-distance between both datasets is 0.000807062
The average-shifted relative 0-distance between both datasets is 3.74353e-10
The absolute 1-distance between both datasets is 0.0147647
The relative 1-distance between both datasets is 2.10156e-12
The average-shifted absolute 1-distance between both datasets is 0.0147742
The average-shifted relative 1-distance between both datasets is 2.10291e-12
The absolute 2-distance between both datasets is 0.00162348
The relative 2-distance between both datasets is 1.50809e-11
The average-shifted absolute 2-distance between both datasets is 0.00162348
The average-shifted relative 2-distance between both datasets is 1.50808e-11

$vlsvdiff_DP --diff Flowthrough_amr_dev_2tasks/bulk.0000001.vlsv Flowthrough_amr_Mpitrans_2tasks/bulk.0000proton/vg_v 0
INFO Reading in two files.
Statistics on file: size 4664 min = 100000 max = 216218 average = 151403 standard deviation 744.591
Statistics on file: size 4664 min = 100000 max = 216218 average = 151403 standard deviation 744.591
The absolute 0-distance between both datasets is 6.63768e-05
The relative 0-distance between both datasets is 3.0699e-10
The average-shifted absolute 0-distance between both datasets is 6.63116e-05
The average-shifted relative 0-distance between both datasets is 3.06689e-10
The absolute 1-distance between both datasets is 0.00185172
The relative 1-distance between both datasets is 2.6223e-12
The average-shifted absolute 1-distance between both datasets is 0.00195854
The average-shifted relative 1-distance between both datasets is 2.77357e-12
The absolute 2-distance between both datasets is 0.000178049
The relative 2-distance between both datasets is 1.6324e-11
The average-shifted absolute 2-distance between both datasets is 0.000177993
The average-shifted relative 2-distance between both datasets is 1.63189e-11

$vlsvdiff_DP --diff Flowthrough_amr_dev_2tasks/bulk.0000001.vlsv Flowthrough_amr_Mpitrans_2tasks/bulk.0000proton/vg_v 1
INFO Reading in two files.
Statistics on file: size 4664 min = -8312.03 max = 3136.24 average = -348.376 standard deviation 28.1942
Statistics on file: size 4664 min = -8312.03 max = 3136.24 average = -348.376 standard deviation 28.1942
The absolute 0-distance between both datasets is 0.000226208
The relative 0-distance between both datasets is 2.72145e-08
The average-shifted absolute 0-distance between both datasets is 0.000226171
The average-shifted relative 0-distance between both datasets is 2.72101e-08
The absolute 1-distance between both datasets is 0.00279288
The relative 1-distance between both datasets is 4.7462e-10
The average-shifted absolute 1-distance between both datasets is 0.00283367
The average-shifted relative 1-distance between both datasets is 4.81551e-10
The absolute 2-distance between both datasets is 0.000416893
The relative 2-distance between both datasets is 3.12035e-09
The average-shifted absolute 2-distance between both datasets is 0.000416886
The average-shifted relative 2-distance between both datasets is 3.12029e-09

$vlsvdiff_DP --diff Flowthrough_amr_dev_2tasks/bulk.0000001.vlsv Flowthrough_amr_Mpitrans_2tasks/bulk.0000proton/vg_v 2
INFO Reading in two files.
Statistics on file: size 4664 min = -8642.66 max = 5819.87 average = -832.077 standard deviation 47.5971
Statistics on file: size 4664 min = -8642.66 max = 5819.87 average = -832.077 standard deviation 47.5971
The absolute 0-distance between both datasets is 0.000141929
The relative 0-distance between both datasets is 1.64219e-08
The average-shifted absolute 0-distance between both datasets is 0.000141851
The average-shifted relative 0-distance between both datasets is 1.64129e-08
The absolute 1-distance between both datasets is 0.00208981
The relative 1-distance between both datasets is 2.2077e-10
The average-shifted absolute 1-distance between both datasets is 0.00223288
The average-shifted relative 1-distance between both datasets is 2.35884e-10
The absolute 2-distance between both datasets is 0.000263682
The relative 2-distance between both datasets is 1.15093e-09
The average-shifted absolute 2-distance between both datasets is 0.000263628
The average-shifted relative 2-distance between both datasets is 1.15069e-09
```
Running some larger performance tests to see if it's worth it

**Pull request closed without merging**